pkg/controller/dynamicupgradeableconditionscontroller.go

```diff
@@ -0,0 +1,245 @@
+	PackageNameKey                 = "olm.operatorframework.io/package-name"
+	BundleNameKey                  = "olm.operatorframework.io/bundle-name"
+	BundleVersionKey               = "olm.operatorframework.io/bundle-version"
+)
+
+type dynamicUpgradeableConditionController struct {
+	name                   string
+	kubeclient             kubernetes.Interface
+	clusterExtensionClient *clients.ClusterExtensionClient
+	operatorClient         *clients.OperatorClient
+	prefixes               []string
+	logger                 logr.Logger
+}
+
+func NewDynamicUpgradeableConditionController(kubeclient kubernetes.Interface, clusterExtensionClient *clients.ClusterExtensionClient, name string, operatorClient *clients.OperatorClient, eventRecorder events.Recorder, prefixes []string) factory.Controller {
+	c := &dynamicUpgradeableConditionController{
+		name:                   name,
+		clusterExtensionClient: clusterExtensionClient,
+		kubeclient:             kubeclient,
+		operatorClient:         operatorClient,
+		prefixes:               prefixes,
+		logger:                 klog.FromContext(context.TODO()).WithName(name),
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).WithInformers(operatorClient.Informer(), clusterExtensionClient.Informer().Informer()).ToController(name, eventRecorder)
+}
+
+func (c *dynamicUpgradeableConditionController) sync(ctx context.Context, _ factory.SyncContext) error {
+	c.logger.V(4).Info("sync started")
+	defer c.logger.V(4).Info("done syncing")
+
+	updateStatusFuncs, err := c.ensureControllersManaged()
+	if err != nil {
+		c.logger.V(4).Error(err, "error verifying OLM state")
+		return err
+	}
+
+	incompatibleOperatorsUpdateStatusFuncs, err := c.checkIncompatibleOperators()
+	if err != nil {
+		c.logger.V(4).Error(err, "error checking incompatible operators")
+		return err
+	}
+
+	updateStatusFuncs = append(updateStatusFuncs, incompatibleOperatorsUpdateStatusFuncs...)
+	if _, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, updateStatusFuncs...); updateErr != nil {
+		c.logger.V(4).Info(fmt.Sprintf("Error updating operator condition status: %v", updateErr))
+		return updateErr
+	}
+	return nil
+}
+
+func (c *dynamicUpgradeableConditionController) ensureControllersManaged() ([]v1helpers.UpdateStatusFunc, error) {
+	c.logger.V(4).Info("verifying state of olm controllers")
+	defer c.logger.V(4).Info("done verifying state of olm controllers")
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return nil, err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil, nil
+	}
+
+	updateStatusFuncs := make([]v1helpers.UpdateStatusFunc, 0, len(c.prefixes))
+	for _, prefix := range c.prefixes {
+		updateStatusFuncs = append(updateStatusFuncs, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   fmt.Sprintf("%sUpgradeable", prefix),
+			Status: operatorv1.ConditionTrue,
+		}))
+	}
+	return updateStatusFuncs, nil
```

Comment:
I don't think we should put these hardcoded `Upgradeable=True` status update functions inside a function called `ensureControllersManaged`. It feels like we're mixing concerns there a bit.

I kinda think we should refactor this slightly to accept a slice map of `[]func(context.Context) (operatorv1.OperatorCondition, error)`, where that function (not sure I have the input parameters correct) returns a Condition that we should set on our ClusterOperator.

If we had that setup, then this controller would boil down to:
1. Check the management state
2. If we're managed, iterate the `[]func(context.Context) (Status, error)`, calling each function to build `updateStatusFuncs`
3. Call `v1helpers.UpdateStatus`.

And it would become a generic `ConditionsController`. And if we had _that_ then we could put the `OLMManagedOperatorsUpgradeable` condition logic into a completely separate package/function that only needs to implement that function signature.